<a href="https://colab.research.google.com/github/billydavisiv/BUAD-427/blob/main/BUAD427_FA26_Homework2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 2: AB testing intro

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


A/B testing (also known as split testing) is a research methodology that involves comparing two versions of a single variable (A and B) to determine which performs better. It is widely used in marketing, web design, and product development to make data-driven decisions. In Python, libraries like `scipy.stats` provide functions for performing the statistical tests necessary to analyze A/B test results, such as t-tests and chi-squared tests.

In [ ]:
import scipy.stats as stats

# The scipy.stats library contains a wide range of statistical functions
# useful for A/B testing, including functions for hypothesis testing
# (e.g., t-tests, chi-squared tests) and probability distributions.

## Example 1: Button1

## Import data

In [ ]:
import pandas as pd

button1 = pd.read_csv('/content/drive/MyDrive/BUAD 427/Week 2/Button_1.csv')

button1.head()

,Customer ID,Purchase Amt,ButtonVersion
0,114,88.52,#1
1,135,94.33,#2
2,149,92.66,#2
3,146,91.34,#2
4,140,93.68,#2


In [ ]:
comparison = button1.groupby('ButtonVersion')['Purchase Amt'].agg(['mean', 'std', 'count'])
print(comparison)

                 mean       std  count
ButtonVersion                         
#1             89.618  3.717436     20
#2             92.350  2.325504     20



### About ttest_ind


The t-test is a crucial statistical tool in A/B testing for determining if observed differences between two group means are statistically significant. In Python, the `scipy.stats` library provides functions like `scipy.stats.ttest_ind` to perform various t-tests, allowing for robust analysis of experimental results by comparing sample means and calculating p-values to assess the evidence against a null hypothesis.

### Key Parameter: `equal_var` in `scipy.stats.ttest_ind`

When performing an independent two-sample t-test using `scipy.stats.ttest_ind`, the `equal_var` parameter is crucial:

*   **`equal_var=True`**: This assumes that the two samples (groups) have equal population variances. This is the assumption for a standard Student's t-test.
*   **`equal_var=False`**: This is the default behavior and performs Welch's t-test. Welch's t-test does *not* assume equal population variances and is generally more robust when the variances are unequal or when you are unsure about their equality. Given that real-world data often has unequal variances, setting `equal_var=False` is a safer and more common practice in A/B testing, as it tends to lower potential Type I errors if the equal variance assumption is violated.

In [ ]:
import scipy.stats as stats

# Separate the purchase amounts for each button version
purchase_amt_v1 = button1[button1['ButtonVersion'] == '#1']['Purchase Amt']
purchase_amt_v2 = button1[button1['ButtonVersion'] == '#2']['Purchase Amt']

# Perform independent t-test (Welch's t-test, assuming unequal variances)
t_statistic, p_value = stats.ttest_ind(purchase_amt_v2, purchase_amt_v1, equal_var=False, alternative='greater')

# Display the results
print(f"T-statistic: {t_statistic:.3f}")
print(f"P-value: {p_value:.3f}")

# Check for statistical significance at 5% level
significance_level = 0.05

if p_value < significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we reject the null hypothesis.")
    print("There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1.")
elif p_value >= significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we fail to reject the null hypothesis.")
    print("There is no significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1.")

T-statistic: 2.786
P-value: 0.004

At the 5% significance level, we reject the null hypothesis.
There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1.


## Example 2: Button2

In [ ]:
import pandas as pd

button2 = pd.read_csv('/content/drive/MyDrive/BUAD 427/Week 2/Button_2.csv')

In [ ]:
comparison_button2 = button2.groupby('ButtonVersion')['Purchase Amt'].agg(['mean', 'std', "count"])
print(comparison_button2)

                  mean        std  count
ButtonVersion                           
#1             88.7765  17.003695     20
#2             93.1965  18.962113     20


In [ ]:
import scipy.stats as stats

# Separate the purchase amounts for each button version from button2 data
purchase_amt_v1_button2 = button2[button2['ButtonVersion'] == '#1']['Purchase Amt']
purchase_amt_v2_button2 = button2[button2['ButtonVersion'] == '#2']['Purchase Amt']

# Perform independent t-test (Welch's t-test, assuming unequal variances)
t_statistic_button2, p_value_button2 = stats.ttest_ind(purchase_amt_v2_button2, purchase_amt_v1_button2, equal_var=False, alternative='greater')

# Display the results
print(f"T-statistic for button2 data: {t_statistic_button2:.3f}")
print(f"P-value for button2 data: {p_value_button2:.3f}")

# Check for statistical significance at 5% level
significance_level = 0.05

if p_value_button2 < significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we reject the null hypothesis for button2 data.")
    print("There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in button2 data.")
elif p_value_button2 >= significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we fail to reject the null hypothesis for button2 data.")
    print("There is no significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in button2 data.")

T-statistic for button2 data: 0.776
P-value for button2 data: 0.221

At the 5% significance level, we fail to reject the null hypothesis for button2 data.
There is no significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in button2 data.


## Example 3: Button 3

In [ ]:
import pandas as pd

button3 = pd.read_csv('/content/drive/MyDrive/BUAD 427/Week 2/Button_3.csv')

button3.head()

,Customer ID,Purchase Amt,ButtonVersion,Region
0,901,84.77,#1,A
1,1887,61.06,#2,B
2,328,41.80,#1,B
3,1783,108.53,#2,A
4,2260,109.45,#2,A


In [ ]:
comparison_button3 = button3.groupby('ButtonVersion')['Purchase Amt'].agg(['mean', 'std', "count"])
print(comparison_button3)

                    mean        std  count
ButtonVersion                             
#1             89.609247  19.269060   1500
#2             91.755300  20.359525   1500


In [ ]:
import scipy.stats as stats

# Separate the purchase amounts for each button version from button3 data
purchase_amt_v1_button3 = button3[button3['ButtonVersion'] == '#1']['Purchase Amt']
purchase_amt_v2_button3 = button3[button3['ButtonVersion'] == '#2']['Purchase Amt']

# Perform independent t-test (Welch's t-test, assuming unequal variances)
t_statistic_button3, p_value_button3 = stats.ttest_ind(purchase_amt_v2_button3, purchase_amt_v1_button3, equal_var=False, alternative='greater')

# Display the results
print(f"T-statistic for button3 data: {t_statistic_button3:.3f}")
print(f"P-value for button3 data: {p_value_button3:.3f}")

# Check for statistical significance at 5% level
significance_level = 0.05

if p_value_button3 < significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we reject the null hypothesis for button3 data.")
    print("There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in button3 data.")
elif p_value_button3 >= significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we fail to reject the null hypothesis for button3 data.")
    print("There is no significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in button3 data.")

T-statistic for button3 data: 2.965
P-value for button3 data: 0.002

At the 5% significance level, we reject the null hypothesis for button3 data.
There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in button3 data.


## **Homework 2**

**Due**

By the end of 01/23/2026

**Assignment Instructions**

* Follow the steps we discussed in class to develop the analysis and answer all questions.

* Write down your prompt either in a new text box or as a line comment in the code box.

* Read the code generated by AI. Focus on the logical flow.

* If necessary, make adjustment and prompt AI to update the code.

**Submission Requirements:**

&nbsp; &nbsp; Convert your completed notebook into a PDF or HTML report.

&nbsp; &nbsp; Submit the report on Brightspace before the deadline.

**Grading Rubric**

Assignments must be complete (all questions answered) and submitted on time. The rubric emphasizes mastery of programming details, as precision is critical in coding.

*  5 – Excellent: Strong effort. Code runs without errors. Explanations align with the results produced.

*  4 – Good: Strong effort. Code contains errors, and some explanations are inconsistent with the results.

*  2 – Fair: Limited effort. Multiple major errors (e.g., incorrect functions, inappropriate features, or results far off expectations). Incomplete work, insufficient explanation, or missing report.

Late Submissions:

*  Up to 24 hours late → 50% of the original grade.

*  More than 24 hours late → 0 points.



### Question 1: In the dataset button3, there is a feature called Region, which records the region where each transaction occurred. How many transactions are in Region A, and how many are in Region B?

In [ ]:
region_counts = button3['Region'].value_counts()
print(region_counts)

Region
A    1504
B    1496
Name: count, dtype: int64


There were 1,504 transactions in Region A and 1,496 transactions in Region B.

### Question 2. For the transactions in Region A, what proportion received ButtonVersion1? What about Region B? Are these proportions comparable? Why is it important to compare these proportions in the experiment?

In [ ]:
total_transactions_A = button3[button3['Region'] == 'A'].shape[0]
button1_transactions_A = button3[(button3['Region'] == 'A') & (button3['ButtonVersion'] == '#1')].shape[0]
proportion_A = button1_transactions_A / total_transactions_A

total_transactions_B = button3[button3['Region'] == 'B'].shape[0]
button1_transactions_B = button3[(button3['Region'] == 'B') & (button3['ButtonVersion'] == '#1')].shape[0]
proportion_B = button1_transactions_B / total_transactions_B

print(f"Proportion of ButtonVersion1 in Region A: {proportion_A:.3f}")
print(f"Proportion of ButtonVersion1 in Region B: {proportion_B:.3f}")

# Analysis of comparability and value of comparison:
# From the results, the proportion of ButtonVersion1 in Region A is 0.501, and in Region B is 0.499.
# These proportions are very close to each other (both approximately 50%).
# This indicates that the allocation of ButtonVersion1 (and by extension, ButtonVersion2) was largely comparable across Region A and Region B.

# Why is it important to compare these proportions in the experiment?
# In A/B testing, it is crucial to ensure that the different groups (in this case, ButtonVersion1 and ButtonVersion2) are exposed to comparable conditions
# across relevant segments (like regions). If the proportion of ButtonVersion1 in Region A was significantly different from that in Region B,
# it would suggest a potential imbalance in the experiment's design or execution.
# Such an imbalance could act as a confounding factor. For example, if Region A had a much higher proportion of ButtonVersion1 and also inherently
# had customers with different purchasing behaviors, any observed differences in purchase amounts between button versions could be mistakenly attributed
# to the button design rather than the regional characteristics or the uneven distribution of button versions.
# By confirming that the proportions are comparable, we increase our confidence that any observed differences in outcomes between button versions
# are truly due to the button versions themselves, and not due to differential exposure across regions.
# This helps maintain the internal validity of the A/B test and ensures more reliable conclusions for business decisions.

Proportion of ButtonVersion1 in Region A: 0.502
Proportion of ButtonVersion1 in Region B: 0.498


We saw that Button Version 1 was exposed to 50.2% of Region A transactions and 49.8% of Region B transactions, which shows us that both button versions received near equal treatment.

This is important because, in AB testing, we want to ensure that both groups are treated under the same conditions. It's like a science experiment in that all conditions outside of an independent variable should be constant.

### Question 3. Find the average purchase amount for Button2 and Button1 in Region A and Region B, respectively.

In [ ]:
import pandas as pd

# Calculate average purchase amounts for Region A
region_A_data = button3[button3['Region'] == 'A']
mean_A_v1 = region_A_data[region_A_data['ButtonVersion'] == '#1']['Purchase Amt'].mean()
mean_A_v2 = region_A_data[region_A_data['ButtonVersion'] == '#2']['Purchase Amt'].mean()

# Calculate average purchase amounts for Region B
region_B_data = button3[button3['Region'] == 'B']
mean_B_v1 = region_B_data[region_B_data['ButtonVersion'] == '#1']['Purchase Amt'].mean()
mean_B_v2 = region_B_data[region_B_data['ButtonVersion'] == '#2']['Purchase Amt'].mean()

# Create a DataFrame to display the results in a 2x2 matrix format
summary_df = pd.DataFrame({
    'ButtonVersion1': [mean_A_v1, mean_B_v1],
    'ButtonVersion2': [mean_A_v2, mean_B_v2]
}, index=['Region A', 'Region B'])

print("Average Purchase Amount by Region and Button Version:")

# Define conditional formatting function
def highlight_max_in_row(s):
    is_max = s == s.max()
    # Set non-max values to white background, max values to green
    return ['background-color: green' if v else 'background-color: gray' for v in is_max]

# Create Styler object
styler = summary_df.style

# Apply conditional formatting to highlight the max value in each row
styler = styler.apply(highlight_max_in_row, axis=1, subset=['ButtonVersion1', 'ButtonVersion2'])

# Display the styled DataFrame
styled_summary_df = styler
styled_summary_df

Average Purchase Amount by Region and Button Version:


,ButtonVersion1,ButtonVersion2
Region A,88.865828,98.596235
Region B,90.362644,84.932583


Using this 2x2 matrix, we can see that there is not a consensus best choice.

For Region A, Button Version 2 (98.60 dollars) led to a higher average purchase amount than Button Version 1 (88.87 dollars).

For Region B, Button Version 1 (90.36 dollars) led to a higher average purchase amount than Button Version 1 (84.93 dollars).

### Question 4. For Region A, repeat the hypotheses as we did in Example 3. At the 5% significance level, does ButtonVersion2 generate significantly higher average revenue than ButtonVersion1?

In [ ]:
import scipy.stats as stats

# Filter data for Region A
region_A_data = button3[button3['Region'] == 'A']

# Separate the purchase amounts for each button version in Region A
purchase_amt_A_v1 = region_A_data[region_A_data['ButtonVersion'] == '#1']['Purchase Amt']
purchase_amt_A_v2 = region_A_data[region_A_data['ButtonVersion'] == '#2']['Purchase Amt']

# Perform independent t-test (Welch's t-test, assuming unequal variances)
t_statistic_A, p_value_A = stats.ttest_ind(purchase_amt_A_v2, purchase_amt_A_v1, equal_var=False, alternative='greater')

# Display the results
print(f"T-statistic for Region A: {t_statistic_A:.3f}")
print(f"P-value for Region A: {p_value_A:.3f}")

# Check for statistical significance at 5% level
significance_level = 0.05

if p_value_A < significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we reject the null hypothesis for Region A.")
    print("There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in Region A.")
elif p_value_A >= significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we fail to reject the null hypothesis for Region A.")
    print("There is no significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in Region A.")

T-statistic for Region A: 10.004
P-value for Region A: 0.000

At the 5% significance level, we reject the null hypothesis for Region A.
There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in Region A.


### Question 5. Repeat the test in the previous question for Region B. Report your finding.

In [ ]:
import scipy.stats as stats

# Filter data for Region B
region_B_data = button3[button3['Region'] == 'B']

# Separate the purchase amounts for each button version in Region B
purchase_amt_B_v1 = region_B_data[region_B_data['ButtonVersion'] == '#1']['Purchase Amt']
purchase_amt_B_v2 = region_B_data[region_B_data['ButtonVersion'] == '#2']['Purchase Amt']

# Perform independent t-test (Welch's t-test, assuming unequal variances)
t_statistic_B, p_value_B = stats.ttest_ind(purchase_amt_B_v2, purchase_amt_B_v1, equal_var=False, alternative='greater')

# Display the results
print(f"T-statistic for Region B: {t_statistic_B:.3f}")
print(f"P-value for Region B: {p_value_B:.3f}")

# Check for statistical significance at 5% level
significance_level = 0.05

if p_value_B < significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we reject the null hypothesis for Region B.")
    print("There is significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in Region B.")
elif p_value_B >= significance_level:
    print(f"\nAt the {significance_level*100:.0f}% significance level, we fail to reject the null hypothesis for Region B.")
    print("There is no significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in Region B.")

T-statistic for Region B: -5.362
P-value for Region B: 1.000

At the 5% significance level, we fail to reject the null hypothesis for Region B.
There is no significant evidence that Button Version #2 generates a higher average purchase amount than Button Version #1 in Region B.


### Question 6. Based on your analysis in the previous two parts, is Region a confounding factor when comparing the average purchase amounts generated by different button versions? How might this influence the business decision? Are there specific situations where it would be inappropriate to customize decisions based on these findings?

In [ ]:
print("Descriptive Statistics for Purchase Amount by Region and Button Version:\n")
button3_analysis = button3.groupby(['Region', 'ButtonVersion'])['Purchase Amt'].agg([
    'mean', 'median', 'std', 'min', 'max', 'count', lambda x: x.mode()[0] if not x.mode().empty else 'N/A'
])
button3_analysis.columns = ['Mean', 'Median', 'Std Dev', 'Min', 'Max', 'Count', 'Mode']

print(button3_analysis)

print("\n\nFurther analysis to consider:\n")
print("1. **Distribution Shape:** Are the distributions skewed? A histogram or box plot for each group (Region-ButtonVersion) could reveal this.")
print("2. **Outliers:** Are there significant outliers driving up/down the average in certain groups? This could be checked with box plots or by examining the min/max values relative to the mean/median.")
print("3. **Customer Demographics (if available):** If there were any demographic data associated with 'Customer ID', we could investigate if certain customer segments (e.g., age, income) are more prevalent in one region and prefer a specific button version.")
print("4. **Interaction Effects:** The t-tests showed that Region *is* a confounding factor. The significant difference in Region A (BV2 > BV1) and no significant difference in Region B (with BV1's mean slightly higher) strongly suggests an interaction effect between Region and ButtonVersion. This means the effect of the button version on purchase amount is not consistent across regions.")
print("5. **Hypotheses:** It's possible that Button Version 2's design elements appeal more to the demographics or cultural preferences prevalent in Region A, while Button Version 1's design elements are more effective in Region B, or vice-versa.")


Descriptive Statistics for Purchase Amount by Region and Button Version:

                           Mean  Median    Std Dev    Min     Max  Count  \
Region ButtonVersion                                                       
A      #1             88.865828   88.32  18.704139  27.39  156.73    755   
       #2             98.596235   98.75  19.015641  30.08  155.83    749   
B      #1             90.362644   90.58  19.809257  30.04  147.25    745   
       #2             84.932583   84.88  19.351744  24.95  151.62    751   

                        Mode  
Region ButtonVersion          
A      #1              62.84  
       #2             109.93  
B      #1              79.74  
       #2              65.52  


Further analysis to consider:

1. **Distribution Shape:** Are the distributions skewed? A histogram or box plot for each group (Region-ButtonVersion) could reveal this.
2. **Outliers:** Are there significant outliers driving up/down the average in certain groups? This could be che

Yes, it absolutely is.

We were able to see that, under identical conditions, that average purchase amount varied to a statistically significant level. With p-values sitting at certain extremes (0.000 for Region A and 1.000 for Region B), we know for certain that it has a confounding effect.

The dataset does not provide enough outside context for us to infer why this is the case, but the answer is likely tied to some sort of demographic or contextual factor. For example, Region A may have been identified as being price-sensitive, so a button that highlighted this may have succeeded here.

Further analysis of the dataset would surely reveal more insights, but this prelimiary analysis proves that Region is a relevant factor in the data.